File: java/scf-simple/src/main/java/org/mydotey/scf/type/string/StringToListConverter.java
package org.mydotey.scf.type.string;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import org.mydotey.scf.type.TypeConverter;

/**
 * @author koqizhao
 *
 * May 21, 2018
 */
public class StringToListConverter<V> extends StringConverter<List<V>> {

    public static final StringToListConverter<String> DEFAULT = new StringToListConverter<>(
            StringInplaceConverter.DEFAULT);

    private TypeConverter<String, V> _typeConverter;

    @SuppressWarnings({ "unchecked", "rawtypes" })
    public StringToListConverter(TypeConverter<String, V> typeConverter) {
        super((Class) List.class);
        Objects.requireNonNull(typeConverter, "typeConverter is null");
        _typeConverter = typeConverter;
    }

    @Override
    public List<V> convert(String source) {
        if (source == null)
            return null;

        source = source.trim();
        if (source.isEmpty())
            return null;

        List<V> list = null;
        String[] array = source.split(",");
        for (String str : array) {
            if (str == null)
                continue;

            str = str.trim();
            if (str.isEmpty())
                continue;

            V v = _typeConverter.convert(str);
            if (v == null)
                continue;

            if (list == null)
                list = new ArrayList<>();

            list.add(v);
        }

        return list;
    }

    @Override
    public int hashCode() {
        final int prime = 31;
        int result = super.hashCode();
        result = prime * result + ((_typeConverter == null) ? 0 : _typeConverter.hashCode());
        return result;
    }

    @SuppressWarnings("rawtypes")
    @Override
    public boolean equals(Object obj) {
        if (!super.equals(obj))
            return false;

        StringToListConverter other = (StringToListConverter) obj;
        return Objects.equals(_typeConverter, other._typeConverter);
    }

    @Override
    public String toString() {
        return String.format("%s { sourceType: %s, targetType: %s, typeConverter: %s }", getClass().getSimpleName(),
                getSourceType(), getTargetType(), _typeConverter);
    }

}

File: dotnet/src/MyDotey.SCF.Simple/Type/String/StringToListConverter.cs
using System;
using System.Collections.Generic;

namespace MyDotey.SCF.Type.String
{
    /**
     * @author koqizhao
     *
     * May 21, 2018
     */
    public class StringToListConverter<V> : StringConverter<List<V>>
    {
        private TypeConverter<string, V> _typeConverter;

        public StringToListConverter(TypeConverter<string, V> typeConverter)
        {
            if (typeConverter == null)
                throw new ArgumentNullException("typeConverter is null");

            _typeConverter = typeConverter;
        }

        public override List<V> Convert(string source)
        {
            if (string.IsNullOrWhiteSpace(source))
                return null;
            source = source.Trim();

            List<V> list = null;
            string[] array = source.Split(',');
            foreach (string s in array)
            {
                if (string.IsNullOrWhiteSpace(s))
                    continue;

                string str = s.Trim();
                V v = _typeConverter.Convert(str);
                if (v == null)
                    continue;

                if (list == null)
                    list = new List<V>();

                list.Add(v);
            }

            return list;
        }

        public override int GetHashCode()
        {
            int prime = 31;
            int result = base.GetHashCode();
            result = prime * result + ((_typeConverter == null) ? 0 : _typeConverter.GetHashCode());
            return result;
        }

        public override bool Equals(Object obj)
        {
            if (!base.Equals(obj))
                return false;

            StringToListConverter<V> other = (StringToListConverter<V>)obj;
            return Object.Equals(_typeConverter, other._typeConverter);
        }

        public override string ToString()
        {
            return string.Format("{0} {{ sourceType: {1}, targetType: {2}, typeConverter: {3} }}", GetType().Name,
                    SourceType, TargetType, _typeConverter);
        }
    }

    public class StringToListConverter : StringToListConverter<string>
    {
        public static readonly StringToListConverter DEFAULT = new StringToListConverter();

        protected StringToListConverter()
            : base(StringInplaceConverter.DEFAULT)
        {

        }
    }
}